### PR TITLE
[FLINK-20695][ha] Clean up ha data for job if globally terminated

### PR DESCRIPTION
**Pull request by yittg**
## What is the purpose of the change

At the moment Flink only cleans up the ha data (e.g. K8s ConfigMaps, or
Zookeeper nodes) when shutting down the cluster. This is not enough for
a long running session cluster to which you submit multiple jobs. In
this change, we clean up the data for a particular job if it reaches a
globally terminal state.

## Brief change log

  - Clean up the ha data for a particular job if it reaches a globally terminal state.

## Verifying this change

This change added tests and can be verified as follows:

  - Added test that validates that the ha data for a particular job is properly deleted.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by yittg** [comment]
@flinkbot run azure

**Comment by yittg** [comment]
@tillrohrmann PTAL.

**Comment by yittg** [comment]
@tillrohrmann  That's absolutely great, it's what i expected while preparing this PR. But will it lead to a breaking change?
And should we hold this PR and wait it be merged?

**Comment by yittg** [comment]
That's good.
And can we apply this PR to the 1.12,1.13 with the old structure?  the 1.14 with the new structure?
I think it's useful to fix it on the old versions, especially the newly released version 1.13.

**Comment by yittg** [comment]
> I see, they are currently being taken care of by the CompletedCheckpointStore and the CheckpointIDCounter when they are being shut down.

I've checked the `/checkpoint-counter` and `/checkpoints` paths for the particular job has be properly deleted.

> Then let's also make sure that we also delete the checkpoints and checkpoint-id-counter paths when calling ZooKeeperHaServices.internalCleanupJobData().

Yes, with the new structure, it's always guaranteed.

**Comment by tillrohrmann** [comment]
Also w/o this change, we should include these paths (`/checkpoint-counter` and `/checkpoints`) to be deleted as well. This makes the `internalCleanupJobData(JobID)` self-contained and independent of the what the other components do before this method is called.

**Comment by yittg** [comment]
Agree, I'll add it later.

**Comment by yittg** [comment]
@tillrohrmann i've add the delete of  `/checkpoint-counter` and `/checkpoints` the `internalCleanupJobData` of the `ZooKeeperHaServices`.

**Review by tillrohrmann — APPROVED** [review]
Really nice work @yittg. Thanks a lot for your contribution! The changes look good to me. I will merge this PR once AZP gives green light. Moreover, I will backport it to `release-1.13` and `release-1.12`.

**Comment by yittg** [comment]
Thanks @tillrohrmann , really appreciate for your help.

**Comment by yittg** [comment]
@tillrohrmann looks like the failed case has nothing to do with this PR? It passed when i run it locally.

**Comment by yittg** [comment]
Hi @tillrohrmann, i just rebased this PR to the newest master to try to pass the AZP.

**Comment by yittg** [comment]
@flinkbot run azure

**Comment by yittg** [comment]
@flinkbot run azure

**Comment by yittg** [comment]
Hi @tillrohrmann , would you merge it ? Just remind you to confirm you didn't miss it, :) thanks again.

**Comment by tillrohrmann** [comment]
Yes, I will merge it now. Sorry, we had a long weekend with a public holiday in Germany so I was a bit out of office.

**Comment by yittg** [comment]
haha, sounds good :). 
Thanks

**Pull request closed without merging**
